src/eval.rs

```diff
@@ -204,14 +204,25 @@ pub fn eval_main<'tcx>(tcx: TyCtxt<'tcx>, main_id: DefId, config: MiriConfig) ->
     // Perform the main execution.
     let res: InterpResult<'_, i64> = (|| {
         // Main loop.
-        while ecx.step()? {
+        loop {
+            match ecx.schedule()? {
+                SchedulingAction::ExecuteStep => {
+                    assert!(ecx.step()?, "a terminated thread was scheduled for execution");
+                }
+                SchedulingAction::ExecuteDtors => {
+                    ecx.run_tls_dtors_for_active_thread()?;
+                }
+                SchedulingAction::Stop => {
+                    break;
+                }
+            }
             ecx.process_diagnostics();
         }
         // Read the return code pointer *before* we run TLS destructors, to assert
         // that it was written to by the time that `start` lang item returned.
         let return_code = ecx.read_scalar(ret_place.into())?.not_undef()?.to_machine_isize(&ecx)?;
         // Global destructors.
-        ecx.run_tls_dtors()?;
+        ecx.run_windows_tls_dtors()?;
```

Comment:
Are we sure the Windows dtor's are global? And there are no per-thread Window dtor's at all? That seems slightly dubious.

Do we have a testcase that checks that per-thread dtors run at the appropriate time, and how does it behave on Windows?

Reply:
> Are we sure the Windows dtor's are global?

No. I have fixed the comment.

> And there are no per-thread Window dtor's at all?

I have added an [assert](https://github.com/rust-lang/miri/pull/1284/files#diff-dc617312c6750224ff59f8b95accca4dR203) that will fail as soon as someone starts implementing concurrency support for Windows, so that they do not forget to also fix the destructors:
```rust
assert_eq!(this.get_total_thread_count()?, 1, "concurrency on Windows not supported");
```

> Do we have a testcase that checks that per-thread dtors run at the appropriate time, and how does it behave on Windows?

We have only [this test](https://github.com/rust-lang/miri/pull/1284/files#diff-b448396480e2fcc4f2dd11ccda8ba9ea) (for Linux and Mac) that checks that the output from the destructor precedes the output from the joining thread.

Reply:
> We have only this test (for Linux and Mac) that checks that the output from the destructor precedes the output from the joining thread.

That sounds like exactly the right kind of test. :)

That is for joined threads. What about destructors of threads that never get joined, in particular detached threads?